streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -35,48 +36,89 @@
 import java.util.HashMap;
-import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
-import java.util.Optional;
-import java.util.Set;
+import java.util.stream.Collectors;
 
 import static java.lang.String.format;
-import static java.util.Collections.emptyMap;
-import static java.util.Collections.unmodifiableList;
 import static org.apache.kafka.streams.processor.internals.StateManagerUtil.CHECKPOINT_FILE_NAME;
 import static org.apache.kafka.streams.processor.internals.StateManagerUtil.converterForStore;
 import static org.apache.kafka.streams.processor.internals.StateRestoreCallbackAdapter.adapt;
 
-
+/**
+ * ProcessorStateManager is the source of truth for the current offset for each state store,
+ * which is either the read offset during restoring, or the written offset during normal processing.
+ *
+ * The offset is initialized as null when the state store is registered, and then it can be updated by
+ * loading checkpoint file, restore state stores, or passing from the record collector's written offsets.
+ *
+ * When checkpointing, if the offset is not null it would be written to the file.
+ *
+ * The manager is also responsible for restoring state stores via their registered restore callback,
+ * which is used for both updating standby tasks as well as restoring active tasks.
+ */
 public class ProcessorStateManager implements StateManager {
+
+    public class StateStoreMetadata {
+        public StateStore stateStore;
+
+        // indicating the current snapshot of the store as the offset of last changelog record that has been
+        // applied to the store used for both restoration (active and standby tasks restored offset) and
+        // normal processing that update stores (written offset); could be null (when initialized)
+        //
+        // the offset is updated in three ways:
+        //   1. when loading from the checkpoint file, when the corresponding task has acquired the state
+        //      directory lock and have registered all the state store; it is only one-time
+        //   2. when updating with restore records (by both restoring active and standby),
+        //      update to the last restore record's offset
+        //   3. when checkpointing with the given written offsets from record collector,
+        //      update blindly with the given offset
+        //
+        // will be used by the changelog reader to determine if restoration is completed, hence public
+        public Long offset;
+
+        // could be used for both active restoration and standby
+        private StateRestoreCallback restoreCallback;
+
+        // record converters used for restoration and standby
+        private RecordConverter recordConverter;
+
+        // corresponding changelog partition of the store
+        private TopicPartition changelogPartition;
+
+        private StateStoreMetadata(final StateStore stateStore) {
+            this.stateStore = stateStore;
+            this.offset = null;
```

Comment:
It's a bit tricky: when you call `seek` or `seekToBeginning` it would not do anything, only the next `poll` would try to do the actual seek. For example if the current log's starting and ending offset is both 10 (i.e. the log is empty):

```
consumer.seek(0);
consumer.position();   <--- return 0;
consumer.poll();          <--- nothing is returned 
consumer.position();   <--- return 10;
```

However:

```
consumer.seekToBeginning();
consumer.position():    <--- return 10; no need to call poll
```

And in our StoreChangelogReader, we only call `position` to trigger the restorer.startRestoration with the starting offset if necessary.

Reply:
I see. Thanks for the info. I'm fine with `null` as a sentinel.